src/code/templates/zoom-challenge.js

```diff
@@ -6,10 +6,47 @@ export default class ZoomChallenge extends Component {
 
   static backgroundClasses = 'fv-layout fv-layout-zoom'
 
+  constructor(props) {
+    super(props);
+    this.state = {
+      hasTutorial: false
+    };
+  }
+
   componentDidMount() {
     var iframeElement = document.getElementById("iframe");
     phone = new iframePhone.ParentEndpoint(iframeElement);
-    phone.addListener('challengeWin', this.props.onWinZoomChallenge);
+    phone.addListener('challengeWin', (activityData) => {
+      let stats;
+      const positiveMetrics = this.props.zoomScoringMetrics;
+
+      // check we have a list of positive metrics and score to calculate
+      if (activityData && activityData.actionStats && positiveMetrics) {
+        stats = activityData.actionStats;
+
+        if (positiveMetrics && stats) {
+          let total = 0, positiveTotal = 0;
+          for (let m in stats) {
+            if (positiveMetrics.indexOf(m) > -1) {
+              positiveTotal += stats[m];
+            }
+            total += stats[m];
+          }
+          // TODO: Calculate actual tolerance for success metric - currently works as:
+          // 90% or higher = blue, 80-90% = yellow, 70-80% = red, etc.
+          const score = positiveTotal / total * 10;
+          const gemScore = 10 - Math.ceil(score);
+          stats.score = gemScore;
+          console.log(stats, score);
+        }
+      }
```

Comment:
My inclination would be to move this block of code into its own function called something `calculateZoomScore()` just to separate the concerns so that `componentDidMount()` is limited to adding the listeners.